### PR TITLE
Resolve N+1 Issue and Add Integration Tests

### PR DESCRIPTION
**Pull request by MarcoDDM**
### Changes Made
- Resolved the N+1 query problem by optimizing database queries.
- Added integration tests for multiple views in the project.

### Checklist
- [x] N+1 query issue resolved.
- [x] Integration tests added and pass successfully.
- [x] Code adheres to coding standards.

### Screenshots (if applicable)
Before fix:
n1-pre

After fix:
n1-posts

RSPEC RESULTS:
rspec_n1

Thanks for your time, and just let me know if you want to make some changes.

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    it 'redirected to users page (users-show)' do
+      visit users_path
+      all_users.each do |user|
+        click_link user.name
+        expect(page).to have_current_path(user_path(user))
+        visit users_path
+      end
+    end
+  end
```
Comment (codepantha) [inline spec/features/user_show_spec.rb]: - [x] For the user show page, the tests which you have written here are quite different from what is expected. Please see the list of the expected tests to be written below 👇 
image

File: spec/features/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  it 'displays the number of likes on a post' do
+    visit user_posts_path(user_id: user.id)
+    expect(page).to have_content(post.likes_counter)
+  end
+
+  it 'redirects to post show page when clicking on post title' do
+    visit user_posts_path(user_id: user.id)
+    click_link post.title
+    expect(page).to have_current_path(user_post_path(user_id: user.id, id: post.id))
+  end
```
Comment (codepantha) [inline spec/features/post_index_spec.rb]: - [x] Please write tests that assert that you  can see a section for pagination 👇 
image

File: spec/features/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  let(:user) do
+    User.create(name: 'Adnan', photo: 'https://unsplash.com/photos/ABC123', posts_counter: 5)
+  end
+
+  let!(:post) do
+    Post.create(title: 'My First Post', text: 'This is my first post', author: user, comments_counter: 0,
+                likes_counter: 0)
+  end
+
+  let!(:comment) do
```
Comment (codepantha) [inline spec/features/posts_index_spec.rb:17]: - [x] Please write featured tests for the user post show page. Kindly see the attached image below for the tests you are expected to write 👇 
image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋,
Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights 🎯 
✔️ Descriptive PR
✔️ Good commit history
✔️ Wrote integration tests 📌 
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,188 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :system do
+  before do
```
Comment (Owusu-Desmond) [inline spec/features/user_show_spec.rb:4]: - [ ] Please note that you are missing a test that checks whether you can view the user's bio. :pray:

File: spec/features/posts_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
+require 'rails_helper'
+
+RSpec.describe 'Post Index', type: :system do
+  before do
```
Comment (Owusu-Desmond) [inline spec/features/posts_index_spec.rb:4]: - [ ] As requested by the previous reviewer, please make sure that you write a Test that asserts you can see a section for pagination if there are more posts than fit on the view. :pray:

**Review by Owusu-Desmond — CHANGES_REQUESTED** [review]
Hi @MarcoDDM , :wave: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some changes you still need to work on before you move on to the next project but you are almost there.



 ## Highlights :+1: 
:heavy_check_mark: Good commit history.
:heavy_check_mark: No linter errors.
:heavy_check_mark: Your project is well documented


 ## Required Changes ♻️
_Check the comments under the review._

- [x] As requested by the previous code reviewer, Please ensure your write featured tests for the user post show page. So I suggest that you create a `post_show_spec.rb` file and write the following test
  image
  

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi team 👋,

Great job implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights 🎯 
✔️ Descriptive PR
✔️ Good commit history
✔️ Wrote integration tests 📌
✔️ No linter errors

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**